catchup/universalFetcher.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package catchup
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/algorand/go-algorand/agreement"
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/network"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/rpcs"
+)
+
+// UniversalFetcher fetches blocks either from an http peer or ws peer.
+type universalFetcher struct {
+	config config.Local
+	net    network.GossipNode
+	log    logging.Logger
+}
+
+// Fetcher queries the current block of the network, and fetches agreed-upon blocks
+type blockFetcher interface {
+	// fetchBlock fetches a block for a given round.
+	fetchBlock(ctx context.Context, round basics.Round, peer network.Peer) (blk *bookkeeping.Block,
+		cert *agreement.Certificate, downloadDuration time.Duration, err error)
+}
+
+// FetcherFactory creates fetchers
+type blockFetcherFactory interface {
+	// Create a new fetcher
+	newBlockFetcher() blockFetcher
+}
+
+type universalBlockFetcherFactory struct {
+	log    logging.Logger
+	net    network.GossipNode
+	config config.Local
+}
+
+func makeUniversalBlockFetcherFactory(log logging.Logger, net network.GossipNode, config config.Local) blockFetcherFactory {
```

Comment:
do we really need this ? it seems like a redundant factory.

Reply:
I introduced it to use the mock fetcher in the block fetcher service tests.

Reply:
Given that it's being used only for testing suggest that the testing are the ones that need to be implemented to changed to adapt to the new code, right ?